### ISSUE TITLE
No refresh_token return?

### ISSUE THREAD

**Issue by lhviet** [google/google-api-nodejs-client#421]
I installed and make all things work well.
Client Side: Google+ Sign In button
Server Side: receive one-time code to exchange...
To get the refresh_token, I add this custom as README suggests:

var url = oauth2Client.generateAuthUrl({
  access_type: 'offline', // 'online' (default) or 'offline' (gets refresh_token)
  scope: scopes // If you only need one scope you can pass it as string
});

However, the result always show only: access_token, "token_type":"Bearer", id_token, and expiry_date.
Do you know how to solve it?

**Comment by LJWall** [comment]
Hi, One thing is that you only get a refresh token on the **first** call for a new account - you're supposed to store it for future use after that.  If you revoke access for your app (either programmatically or go to your google account settings / connected apps), do you get a refresh token on the first call?

**Comment by lhviet** [comment]
Of course, I revoke the permission for my web in here: https://security.google.com/settings/security/permissions

Then, in the interface, this is the Google+ Sign In button:
function render() {
          gapi.signin.render('google_signinButton', {
            'callback': 'googleSigninCallback',
            'responsetype': 'code',
            'accesstype': 'offline',
            'granttype': 'authorization_code',
            'clientid': '1085077170680-p1fudmtj7vc1pg78st9frc9efud5s0bm.apps.googleusercontent.com',
            'redirecturi': 'postmessage',
            'cookiepolicy': 'single_host_origin',
            'requestvisibleactions': 'http://schema.org/AddAction',
            'scope': 'https://www.googleapis.com/auth/plus.login email profile'
          });
        }
The code in client side then send to my server (nodejs): 4/ZdkeZiILo6TXMLgyKR4L8_YyFSltEfLJEE68NNdagGE.AgMGz5Gn6FcSBrG_bnfDxpJWfe5PmgI

Certainly, it still cannot return the refresh_token in server side.

In client side, I got this kind of error (my site is http protocol): Uncaught SecurityError: Blocked a frame with origin "http://bibooki.com" from accessing a frame with origin "https://accounts.google.com".  The frame requesting access has a protocol of "http", the frame being accessed has a protocol of "https". Protocols must match.

Do you think that is the reason I cannot get the refresh_token?

**Comment by LJWall** [comment]
If the code is getting to your nodejs server then maybe the http/https issue is separate.  I can put up a complete working example later perhaps..  I could be wrong, but I don't think you need to generate a redirect url with `generateAuthUrl` if you're using `gapi` on the client side to render a login button?

**Comment by LJWall** [comment]
Hi Again, here's a working example: https://github.com/LJWall/google-auth2-example .

It's not doing much, but it does successfully get a refresh token when access is first granted..

**Comment by lhviet** [comment]
Hi, thank you first.

It's strange. With your code, I can obtain the refresh_token now. Maybe the official GOOGLE+ Sign-in button does not allows for offline access with refresh_token because the authResult responded has different content with the one by your code.
Anyway, it works, thank you very much :+1:

**Comment by Rosensweig** [comment]
I'm looking at this long after the fact, but in case anyone else is reading this thread for your own issues, it looks like lhviet's original problem was because you did accesstype instead of access_type.

My problem, which I was also able to solve with the info here, was that I hadn't saved the refresh token the first time. I added the access_type:"offline" parameter later, and Google was not providing it for users that had logged in previously. Once I revoked access, and logged in again, I got the refresh token with everything else.
